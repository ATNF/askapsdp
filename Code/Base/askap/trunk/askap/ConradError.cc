//#  ConradError.cc: Base class for CONRAD exceptions
//#
//#  Copyright (C) 2007
//#
//#  $Id$

#include <askap/ConradError.h>

namespace conrad {

  ConradError::ConradError(const std::string& message)
    : std::runtime_error(message)
  {}

  ConradError::~ConradError() throw()
  {}


  CheckError::CheckError(const std::string& message)
    : ConradError(message)
  {}

  CheckError::~CheckError() throw()
  {}


  AssertError::AssertError(const std::string& message)
    : ConradError(message)
  {}

  AssertError::~AssertError() throw()
  {}

} // end namespaces
