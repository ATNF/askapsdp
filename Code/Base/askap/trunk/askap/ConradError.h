/// @file
/// @brief Base class for CONRAD exceptions
/// @author Ger van Diepen (gvd AT astron DOT nl)

#ifndef CONRAD_ERROR_H
#define CONRAD_ERROR_H

#include <string>
#include <ostream>
#include <sstream>
#include <stdexcept>

namespace conrad {

  /// Define the base conrad exception class.
  class ConradError: public std::runtime_error
  {
  public:
    /// Constructor taking a message
    explicit ConradError(const std::string& message);
    /// empty destructor
    virtual ~ConradError() throw();
  };

  /// Define the conrad check exception class.
  class CheckError: public ConradError
  {
  public:
    /// Constructor taking a message
    explicit CheckError(const std::string& message);
    /// empty destructor
    virtual ~CheckError() throw();
  };

  /// Define the conrad assert exception class.
  class AssertError: public ConradError
  {
  public:
    /// Constructor taking a message
    explicit AssertError(const std::string& message);
    /// empty destructor
    virtual ~AssertError() throw();
  };


  /// Macro to throw an exception where the exception text can be formatted.
  /// For example:
  /// CONRADTHROW(ConradError, "File " << fileName << " could not be opened");
  //  Note that a do/while is used to be able to use the macro everywhere.
#define CONRADTHROW(exc,messageStream)   \
    {					 \
      std::ostringstream conrad_log_oss; \
      conrad_log_oss << messageStream <<" (thrown in "<<__FILE__  \
		   << ':' << __LINE__<<") ";	 \
      throw exc(conrad_log_oss.str());   \
    }

  /// Check a condition and throw an exception if it fails.
  /// It appends the confdition text to the message stream.
#define CONRADCHECK(condition,messageStream) \
    if (!(condition)) {			     \
      CONRADTHROW(CheckError, messageStream \
                   << " ('" << #condition << "' failed)");	\
    }

  /// Do an assert and throw an exception with the file and line if it fails.
#define CONRADASSERT(condition) \
    if (!(condition)) {	\
      CONRADTHROW(AssertError, #condition << " failed");  \
    }

  /// Do an assert only if in debug mode.
#ifdef CONRAD_DEBUG
#define CONRADDEBUGASSERT(condition) CONRADASSERT(condition)
#else
#define CONRADDEBUGASSERT(condition)
#endif


} // end namespaces

#endif
