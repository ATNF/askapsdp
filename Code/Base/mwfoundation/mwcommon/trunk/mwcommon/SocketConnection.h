/// @file
/// @brief Connection to workers based on a socket.
///
/// @copyright (c) 2007 CSIRO
/// Australia Telescope National Facility (ATNF)
/// Commonwealth Scientific and Industrial Research Organisation (CSIRO)
/// PO Box 76, Epping NSW 1710, Australia
/// atnf-enquiries@csiro.au
///
/// This file is part of the ASKAP software distribution.
///
/// The ASKAP software distribution is free software: you can redistribute it
/// and/or modify it under the terms of the GNU General Public License as
/// published by the Free Software Foundation; either version 2 of the License,
/// or (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program; if not, write to the Free Software
/// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
///
/// @author Ger van Diepen (diepen AT astron nl)
///
//# $Id$

#ifndef ASKAP_MWCOMMON_SOCKETCONNECTION_H
#define ASKAP_MWCOMMON_SOCKETCONNECTION_H

#include <mwcommon/MWConnection.h>
#include <Common/Net/Socket.h>
#include <boost/shared_ptr.hpp>
#include <string>

namespace askap { namespace mwbase {

  /// @ingroup mwcommon
  /// @brief Connection to workers based on a socket.

  /// This class handles the socket connection between two processes.
  /// For a client it can set up the connection to a server on a given
  /// host and port.
  /// For a server it can hold the connection created by SocketListener.
  ///
  /// It is meant to send and receive blobs. The length of a message to
  /// receive is read (by base class MWConnection) from the blob header.

  class SocketConnection: public MWConnection
  {
  public:
    /// Define a shared pointer to this object.
    typedef boost::shared_ptr<SocketConnection> ShPtr;

    /// Set up the client side of a connection.
    /// Upon the first send or receive it connects to the server
    /// on the given host and port.
    /// If the making the connection fails, it will sleep one second and try
    /// again for up to 60 attempts. In this way the case is handled
    /// where a server is started a bit later than a client.
    SocketConnection (const std::string& hostName, const std::string& port);

    /// Add a socket from the server when it accepted a connection
    /// (used by SocketListener).
    /// It takes over the ownership of the pointer.
    explicit SocketConnection (LOFAR::Socket*);

    virtual ~SocketConnection();

    /// Check the state of the connection.
    virtual bool isConnected() const;

    /// Get the length of the message.
    /// Always returns -1 indicating that the length has to be read
    /// from the header.
    virtual int getMessageLength();

    /// Receive the data sent by the destination
    /// and wait until data has been received into buf.
    virtual void receive (void* buf, unsigned size);

    /// Send the data to the destination
    /// and wait until the data has been sent.
    virtual void send (const void* buf, unsigned size);

    /// Get the name of the host this process is running on.
    /// If sockets are not supported (e.g. Cray), it returns an empty string.
    static std::string getHostName();

  private:
    /// Initialize the connection.
    void init();

    LOFAR::Socket  itsConnSocket;
    LOFAR::Socket* itsDataSocket;
  };

}} /// end namespaces

#endif
