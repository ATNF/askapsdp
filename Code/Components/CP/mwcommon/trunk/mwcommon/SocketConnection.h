/// @file
/// @brief Connection to workers based on a socket.
///
/// @copyright (c) 2007 CONRAD, All Rights Reserved.
/// @author Ger van Diepen (diepen AT astron nl)
///
//# $Id$

#ifndef CONRAD_MWCOMMON_SOCKETCONNECTION_H
#define CONRAD_MWCOMMON_SOCKETCONNECTION_H

#include <mwcommon/MWConnection.h>
#include <Common/Net/Socket.h>
#include <boost/shared_ptr.hpp>
#include <string>

namespace conrad { namespace cp {

  /// @ingroup mwcommon
  /// @brief Connection to workers based on a socket.

  /// This class handles the socket connection between two processes.
  /// For a client it can set up the connection to a server on a given
  /// host and port.
  /// For a server it can hold the connection created by SocketListener.
  ///
  /// It is meant to send and receive blobs. The length of a message to
  /// receive is read (by base class MWConnection) from the blob header.

  class SocketConnection: public MWConnection
  {
  public:
    /// Define a shared pointer to this object.
    typedef boost::shared_ptr<SocketConnection> ShPtr;

    /// Set up the client side of a connection.
    /// Upon the first send or receive it connects to the server
    /// on the given host and port.
    /// If the making the connection fails, it will sleep one second and try
    /// again for up to 60 attempts. In this way the case is handled
    /// where a server is started a bit later than a client.
    SocketConnection (const std::string& hostName, const std::string& port);

    /// Add a socket from the server when it accepted a connection
    /// (used by SocketListener).
    /// It takes over the ownership of the pointer.
    explicit SocketConnection (LOFAR::Socket*);

    virtual ~SocketConnection();

    /// Check the state of the connection.
    virtual bool isConnected() const;

    /// Get the length of the message.
    /// Always returns -1 indicating that the length has to be read
    /// from the header.
    virtual int getMessageLength();

    /// Receive the data sent by the destination
    /// and wait until data has been received into buf.
    virtual void receive (void* buf, unsigned size);

    /// Send the data to the destination
    /// and wait until the data has been sent.
    virtual void send (const void* buf, unsigned size);

    /// Get the name of the host this process is running on.
    /// If sockets are not supported (e.g. Cray), it returns an empty string.
    static std::string getHostName();

  private:
    /// Initialize the connection.
    void init();

    LOFAR::Socket  itsConnSocket;
    LOFAR::Socket* itsDataSocket;
  };

}} /// end namespaces

#endif
