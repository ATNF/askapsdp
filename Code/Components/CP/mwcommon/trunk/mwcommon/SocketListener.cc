//# SocketListener.cc: Class that creates a socket and accepts connections
//#
//# @copyright (c) 2007 CONRAD, All Rights Reserved.
//# @author Ger van Diepen <diepen AT astron nl>
//#
//# $Id$

#include <mwcommon/SocketListener.h>
#include <mwcommon/MWError.h>


namespace conrad { namespace cp {

  SocketListener::SocketListener (const std::string& port)
    : itsConnSocket (new LOFAR::Socket("mwsck", port))
  {}

  SocketConnection::ShPtr SocketListener::accept()
  {
    LOFAR::Socket* socket = itsConnSocket->accept();
    SocketConnection::ShPtr dataConn(new SocketConnection(socket));
    int status = itsConnSocket->errcode();
    CONRADCHECK (socket  &&  status == LOFAR::Socket::SK_OK,
                 "SocketConnection server did not accept on host "
                 << itsConnSocket->host() << ", port " << itsConnSocket->port()
                 << ", LOFAR::Socket status " << status << ' '
                 << itsConnSocket->errstr());
    CONRADASSERT (dataConn->isConnected());
    return dataConn;
  }

}} // end namespaces
