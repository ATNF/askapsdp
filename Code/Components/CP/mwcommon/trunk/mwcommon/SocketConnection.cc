//# SocketConnection.cc: Connection to workers based on a socket
//#
//# @copyright (c) 2007 ASKAP, All Rights Reserved.
//# @author Ger van Diepen <diepen AT astron nl>
//#
//# $Id$

#include <mwcommon/SocketConnection.h>
#include <mwcommon/MWError.h>
#include <unistd.h>             // for gethostname


namespace askap { namespace cp {

  SocketConnection::SocketConnection (const std::string& hostName,
                                      const std::string& port)
    : itsConnSocket ("mwsck", hostName, port),
      itsDataSocket (0)
  {}

  SocketConnection::SocketConnection (LOFAR::Socket* conn)
    : itsDataSocket (conn)
  {}

  SocketConnection::~SocketConnection()
  {
    if (itsDataSocket != &itsConnSocket) {
      delete itsDataSocket;
    }
  }

  bool SocketConnection::isConnected() const
  {
    return itsDataSocket  &&  itsDataSocket->isConnected();
  }

  int SocketConnection::getMessageLength()
  {
    return -1;
  }

  void SocketConnection::receive (void* buf, unsigned size)
  {
    if (!itsDataSocket) {
      init();
    }
    char* cbuf = static_cast<char*>(buf);
    while (size > 0) {
      int sz = itsDataSocket->read (cbuf, size);
      ASKAPCHECK (sz>=0,
                   "Read on socket failed: " << itsDataSocket->errstr());
      cbuf += sz;
      size -= sz;
    }
  }

  void SocketConnection::send (const void* buf, unsigned size)
  {
    if (!itsDataSocket) {
      init();
    }
    itsDataSocket->writeBlocking (buf, size);
  }

  void SocketConnection::init()
  {
    // Create a client socket.
    // Try to connect: may fail if no listener started yet.
    // So retry during one minute.
    int status;
    for (int i=0; i<60; ++i) {
      status = itsConnSocket.connect();
      if (status == LOFAR::Socket::SK_OK) {
        // Connected, so socket can be used to send/receive data.
        itsDataSocket = &itsConnSocket;
        break;
      }
      sleep (1);
    }
    ASKAPCHECK (status == LOFAR::Socket::SK_OK,
                 "SocketConnection client could not connect to host "
                 << itsConnSocket.host() << ", port "
                 << itsConnSocket.port()
                 << ", LOFAR::Socket status " << status << ' '
                 << itsConnSocket.errstr());
    ASKAPASSERT (isConnected());
  }

  std::string SocketConnection::getHostName()
  {
    char nm[256];
    ::gethostname(nm, sizeof(nm));
    return std::string(nm);
  }

}} // end namespaces
