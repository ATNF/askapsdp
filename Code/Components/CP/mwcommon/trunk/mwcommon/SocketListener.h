/// @file
/// @brief Class that creates a socket and accepts connections.
///
/// @copyright (c) 2007 CSIRO
/// Australia Telescope National Facility (ATNF)
/// Commonwealth Scientific and Industrial Research Organisation (CSIRO)
/// PO Box 76, Epping NSW 1710, Australia
/// atnf-enquiries@csiro.au
///
/// This file is part of the ASKAP software distribution.
///
/// The ASKAP software distribution is free software: you can redistribute it
/// and/or modify it under the terms of the GNU General Public License as
/// published by the Free Software Foundation; either version 2 of the License,
/// or (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program; if not, write to the Free Software
/// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
///
/// @author Ger van Diepen (diepen AT astron nl)
///
//# $Id$

#ifndef ASKAP_MWCOMMON_SOCKETLISTENER_H
#define ASKAP_MWCOMMON_SOCKETLISTENER_H

#include <mwcommon/SocketConnection.h>
#include <Common/Net/Socket.h>
#include <boost/shared_ptr.hpp>
#include <string>

namespace askap { namespace cp {

  /// @ingroup mwcommon
  /// @brief Class that creates a socket and accepts connections.

  /// This class sets up a socket listener. It is used by SocketConnectionSet
  /// to accept connection requests from workers.
  ///
  /// Internally the class uses a shared pointer to a socket object.
  /// It means that a copy of a SocketListener object can be made, but that
  /// copies share the same underlying socket object.

  class SocketListener
  {
  public:
    /// Set up the server side of a listener.
    explicit SocketListener (const std::string& port);

    /// Listen to a connection and accept it.
    /// It blocks until another process wants to connect.
    SocketConnection::ShPtr accept();

  private:
    boost::shared_ptr<LOFAR::Socket> itsConnSocket;
  };

}} /// end namespaces

#endif
