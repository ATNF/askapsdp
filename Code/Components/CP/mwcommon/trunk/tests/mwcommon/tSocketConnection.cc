//#  tSocketConnection.cc: Program to test SocketConnection
//#
//#  Copyright (C) 2007
//#
//#  $Id$

#include <mwcommon/SocketConnection.h>
#include <mwcommon/SocketListener.h>
#include <mwcommon/ConradError.h>
#include <iostream>
#include <unistd.h>

using namespace conrad;
using namespace conrad::cp;
using namespace std;

void doClient (const string& host, const string& port)
{
  //  sleep (2);
  cout << "Client connection on host " << host
       << ", port " << port << endl;
  SocketConnection socket(host, port);
  double dv = 1;
  socket.send (&dv, sizeof(dv));
  cout << "sent " << dv << endl;
  float fv = 0;
  socket.receive (&fv, sizeof(fv));
  CONRADASSERT (fv == 2);
  cout << "received " << fv << endl;
  sleep(2);
  socket.receive (&fv, sizeof(fv));
  CONRADASSERT (fv == 3);
  cout << "received " << fv << endl;
  dv = 2;
  socket.send (&dv, sizeof(dv));
  cout << "sent " << dv << endl;
}

void doServer (const string& port)
{
  cout << "Server connection on port " << port << endl;
  SocketListener listener(port);
  SocketConnection::ShPtr socket = listener.accept();
  double dv = 0;
  socket->receive (&dv, sizeof(dv));
  CONRADASSERT (dv == 1);
  cout << "received " << dv << endl;
  float fv = 2;
  socket->send (&fv, sizeof(fv));
  cout << "sent " << fv << endl;
  fv = 3;
  socket->send (&fv, sizeof(fv));
  cout << "sent " << fv << endl;
  socket->receive (&dv, sizeof(dv));
  CONRADASSERT (dv == 2);
  cout << "received " << dv << endl;
}

int main(int argc, const char* argv[])
{
  int status = 0;
  try {
    if (argc < 2) {
      cerr << "Run as:" << endl;
      cerr << "  as server:    tSocketConnection port" << endl;
      cerr << "  as client:    tSocketConnection port host" << endl;
    } else if (argc == 2) {
      doServer (argv[1]);
    } else {
      doClient (argv[2], argv[1]);
    }
  } catch (std::exception& x) {
    cout << "Unexpected exception in " << argv[0] << ": " << x.what() << endl;
    status = 1;
  }
  exit(status);
}
