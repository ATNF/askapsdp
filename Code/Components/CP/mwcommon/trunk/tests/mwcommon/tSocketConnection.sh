#!/bin/sh
$lofar_sharedir/runtest.sh tSocketConnection > tSocketConnection.log 2>&1
