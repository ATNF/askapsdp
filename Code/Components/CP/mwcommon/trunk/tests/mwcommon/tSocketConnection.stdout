Run1 ...
Server connection on port 3851
received 1
sent 2
sent 3
received 2
Client connection on host localhost, port 3851
sent 1
received 2
received 3
sent 2
Run2 ...
Server connection on port 3851
received 1
sent 2
sent 3
received 2
Client connection on host localhost, port 3851
sent 1
received 2
received 3
sent 2
